lib/logger.js

```diff
@@ -1,456 +1,634 @@
+
+        if (retrying) {
+          this[kIsLoggingBackedOff] = true
+          this[kBackoffMillis] = backoffWithJitter(
+            this.baseBackoffMillis
+          , this.maxBackoffMillis
+          , this[kBackoffMillis]
+          )
+          setTimeout(() => {
+            this.send(false)
+          }, this[kBackoffMillis])
+          return
+        }
 
-  if (this._buf.length === 0) {
-    debug('Nothing to flush')
-    return callback()
-  }
+        // User-level errors will be discarded since they will never succeed
+        this[kIsSending] = false
+        this[kTotalLinesReady] -= buffer.length
+        this[kAttempts] = 0
+        this[kReadyToSend].shift()
+        buffer.length = 0
+        if (this[kReadyToSend].length) {
+          this.send()
+          return
+        }
 
-  const data = stringify({e: 'ls', ls: this._buf})
-  this._req.qs.now = Date.now()
-
-  this.__dbgLines = this._buf.map(function(msg) { return msg.line })
-  this.__bufSizePreserve = this._bufByteLength
-  this.callback = callback
-  this._bufByteLength = 0
-  this._buf.length = 0
-
-  const _config = {
-    method: 'post'
-  , url: this._url + '?' + querystring.stringify(this._req.qs)
-  , headers: this._req.headers
-  , data
-  , timeout: this._req.timeout
-  , withCredentials: this._req.withCredentials
-  , json: true
-  }
-  if (this._req.useHttps) {
-    _config.httpsAgent = this._req.agent
-  } else {
-    _config.httpAgent = this._req.agent
+        this.emit('cleared', {
+          message: ALL_CLEAR_SENT
+        })
+      })
   }
-
-  this._sendRequest(_config, this)
 }
 
 /*
  *  Populate short-hand for each supported Log Level
  */
-configs.LOG_LEVELS.forEach(function(level) {
-  const levelFunctionName = level.toLowerCase()
-  Logger.prototype[levelFunctionName] = function(statement, opts, callback) {
-    if (opts && typeof opts === 'function') {
-      callback = opts
-      opts = {}
-    }
-    opts = opts || {}
-    opts.level = level
-
-    this.log(statement, opts, callback)
-  }
-})
-
-const flushAll = function(cb) {
-  if (!cb || typeof cb !== 'function') {
-    cb = debug
-  }
-
-  // Variables used to wait for all the flushes to be completed
-  let counter = 0
-  const loggersCount = loggers.length
-  const complete = (errors) => {
-    if (errors.length > 0) {
-      return cb(`The following errors happened while flushing all loggers: ${errors}`)
-    }
-    return cb()
-  }
-
-  const errors = []
-  loggers.forEach((logger) => {
-    logger._flush((err) => {
-      if (err) {
-        errors.push(err)
-      }
-      // Counter is incremented after each flush
-      if (++counter === loggersCount) {
-
-        /*
-         * Flushed all the loggers
-         * Completes by calling the callback
-         */
-        complete(errors)
+for (const level of constants.LOG_LEVELS) {
+  const levelLowerCase = level.toLowerCase()
+
+  Object.defineProperties(Logger.prototype, {
+    [levelLowerCase]: {
+      enumerable: true
```

Comment:
Should probably make the documented methods in the README enumerable as well (they are not by default).